File: src/csharp/Sessionless/Models/ISessionless.cs
﻿using Org.BouncyCastle.Crypto.Parameters;
using SessionlessNET.Impl;

namespace SessionlessNET.Models;

/// <summary> 
/// <para> Passwordless authentication </para>
/// <para> With a <see cref="Vault"/> for storing and retrieving <see cref="KeyPairHex"/>es </para>
/// </summary>
public interface ISessionless {
    /// <summary> The way to store and retrieve <see cref="KeyPairHex"/>s </summary>
    public IVault Vault { get; }

    /// <summary> Creates a unique UUID for a user. </summary> 
    public string GenerateUUID();

    /// <summary>
    /// Generates a private/public <see cref="KeyPairHex"/> and stores it using <paramref name="Vault"/>.
    /// </summary>
    /// <returns> <see cref="KeyPairHex"/> that was generated </returns>
    public KeyPairHex GenerateKeys();

    /// <summary>
    /// Generates a private/public <see cref="KeyPairHex"/> asynchronously and stores it using <paramref name="Vault"/>.
    /// </summary>
    /// <returns> <see cref="KeyPairHex"/> that was generated </returns>
    public Task<KeyPairHex> GenerateKeysAsync();

    /// <summary>
    /// Retrieves keys using <paramref name="Vault"/>.
    /// </summary>
    /// <returns>Key pair as <see cref="KeyPairHex"/>.</returns>
    public KeyPairHex? GetKeys();

    /// <summary>
    /// Signs a <paramref name="message"/> with the user's stored private key
    /// <br/> (Get from <paramref name="Vault"/> using <see cref="GetKeys"/>).
    /// </summary>
    /// <param name="message">The message to be signed.</param>
    /// <returns>Signature as a <see cref="MessageSignatureHex"/>.</returns>
    public MessageSignatureHex Sign(string message);

    /// <summary>
    /// Signs a <paramref name="message"/> using the provided <paramref name="privateKeyHex"/>
    /// </summary>
    /// <param name="message">The message to be signed.</param>
    /// <param name="privateKeyHex"> The private key in hex format to use for signing </param>
    /// <returns>Signature as a <see cref="MessageSignatureHex"/>.</returns>

    public MessageSignatureHex Sign(string message, string privateKeyHex);

    /// <summary>
    /// Signs a <paramref name="message"/> using the provided <paramref name="privateKey"/>.
    /// </summary>
    /// <param name="message">The message to be signed.</param>
    /// <param name="privateKey">The private key to use for signing.</param>
    /// <returns>Signature as a <see cref="MessageSignatureHex"/>.</returns>
    public MessageSignatureHex Sign(string message, ECPrivateKeyParameters privateKey);

    /// <summary>
    /// Verifies a given <paramref name="signedMessage"/> with the user's stored public key.
    /// <br/>
    /// <br/> See: <seealso cref="Sign(string)"/>
    /// </summary>
    /// <param name="signedMessage"> The message that was signed earlier </param>
    /// <returns> True if the signature is valid for the given message and public key </returns>
    public bool VerifySignature(SignedMessage signedMessage);
    /// <summary>
    /// Verifies a given <paramref name="signedMessage"/> with the included <see cref="SignedMessageWithKey.PublicKey"/>.
    /// <br/>
    /// <br/> See: <seealso cref="Sign(string, string)"/>
    /// </summary>
    /// <param name="signedMessage">The message that was signed earlier </param>
    /// <returns> True if the signature is valid for the given message and public key </returns>
    public bool VerifySignature(SignedMessageWithKey signedMessage);
    /// <summary>
    /// Verifies a given <paramref name="signedMessage"/> with the included <see cref="SignedMessageWithECKey.PublicKey"/>
    /// <br/>
    /// <br/> See: <seealso cref="Sign(string, ECPrivateKeyParameters)"/>
    /// </summary>
    /// <param name="signedMessage">The message that was signed earlier </param>
    /// <returns> True if the signature is valid for the given message and public key </returns>
    public bool VerifySignature(SignedMessageWithECKey signedMessage);

    /// <summary>
    /// Verifies each of the <paramref name="messages"/>
    /// <br/>
    /// <br/> See: <seealso cref="VerifySignature"/>
    /// </summary>
    /// <param name="messages"> Messages to be verified </param>
    /// <returns> True if all signatures were verified successfully </returns>
    /// <exception cref="ArgumentException"></exception> 
    public bool Associate(params SignedMessage[] messages);
}

File: src/kotlin/standalone/sessionless/src/main/kotlin/com/planetnine/sessionless/models/ISessionless.kt
package com.planetnine.sessionless.models

import com.planetnine.sessionless.impl.IMessageSignature
import com.planetnine.sessionless.impl.KeyAccessInfo
import com.planetnine.sessionless.impl.KeyPairHex
import com.planetnine.sessionless.impl.KeyStoreVault
import com.planetnine.sessionless.impl.SignedMessage
import com.planetnine.sessionless.impl.SignedMessageWithECKey
import com.planetnine.sessionless.impl.SignedMessageWithKey
import com.planetnine.sessionless.models.vaults.ICustomVault
import com.planetnine.sessionless.models.vaults.IVault
import java.security.KeyPair
import java.security.PrivateKey

interface ISessionless {
    /** The way to store and retrieve key pairs */
    val vault: IVault

    interface WithKeyStore : ISessionless {
        override val vault: KeyStoreVault

        /** Generates a private/public key pair and stores it using [vault]
         * @param keyAccessInfo Info required to access the stored key
         * @return The [KeyPair] that was generated then stored with [vault] */
        fun generateKeys(keyAccessInfo: KeyAccessInfo): KeyPair

        /** Generates a private/public key pair and stores it using [vault]
         * @param keyAccessInfo Info required to access the stored key
         * @return The [KeyPair] that was generated then stored with [vault] */
        suspend fun generateKeysAsync(keyAccessInfo: KeyAccessInfo): KeyPair

        /** Retrieves keys using [vault]
         * @param keyAccessInfo Info required to access the stored key
         * @return Key pair as objects ([KeyPair]) */
        fun getKeys(keyAccessInfo: KeyAccessInfo): KeyPair

        /** Signs a [message] with the user's stored private key (from [vault])
         * @param message The message to be signed.
         * @param keyAccessInfo Info required to access the stored key
         * @return Signature as [IMessageSignature] */
        fun sign(message: String, keyAccessInfo: KeyAccessInfo): IMessageSignature


        /** Verifies a given [signedMessage] with the user's stored public key.
         * See: [sign]
         * @param signedMessage The message that was signed earlier
         * @return True if the signature is valid for the given message and public key */
        fun verifySignature(signedMessage: SignedMessage, keyAccessInfo: KeyAccessInfo): Boolean
    }

    interface WithCustomVault : ISessionless {
        override val vault: ICustomVault

        /** Generates a private/public key pair and stores it using [vault]
         * @return A simplified ([KeyPairHex]) version of the key pair that was generated then stored with [vault] */
        fun generateKeys(): KeyPairHex

        /** Generates a private/public key pair and stores it using [vault]
         * @return A simplified ([KeyPairHex]) version of the key pair that was generated then stored with [vault] */
        suspend fun generateKeysAsync(): KeyPairHex

        /** Retrieves keys using [vault]
         * @return Key pair as [String]s ([KeyPairHex]). */
        fun getKeys(): KeyPairHex?

        /** Signs a [message] with the user's stored private key (from [vault])
         * @param message The message to be signed
         * @return Signature as [IMessageSignature] */
        fun sign(message: String): IMessageSignature

        /** Verifies a given [signedMessage] with the user's stored public key
         * @param signedMessage The message that was signed earlier
         * @return True if the signature is valid for the given message and public key */
        fun verifySignature(signedMessage: SignedMessage): Boolean
    }

    /** Signs a [message] using the provided [privateKey]
     * @param message The message to be signed
     * @return Signature as [IMessageSignature]. */
    fun sign(message: String, privateKey: PrivateKey): IMessageSignature


    /** Verifies a given [signedMessage] with the included [SignedMessageWithKey.publicKey]
     * @param signedMessage The message that was signed earlier
     * @return True if the signature is valid for the given message and public key */
    fun verifySignature(signedMessage: SignedMessageWithKey): Boolean

    /** Verifies a given [signedMessage] with the included [SignedMessageWithECKey.publicKey]
     * @param signedMessage The message that was signed earlier
     * @return True if the signature is valid for the given message and public key */
    fun verifySignature(signedMessage: SignedMessageWithECKey): Boolean


    /** Creates a unique UUID for a user
     * @return The generated UUID. */
    fun generateUUID(): String

    /** Associates [signedMessages] signatures with their respective public keys
     * @return true if all were successfully verified
     * @throws java.lang.IllegalArgumentException if array size<2
     * @see verifySignature */
    fun associate(vararg signedMessages: SignedMessageWithKey): Boolean

    /** Associates [signedMessages] signatures with their respective public keys
     * @return true if all were successfully verified
     * @throws java.lang.IllegalArgumentException if array size<2
     * @see verifySignature */
    fun associate(vararg signedMessages: SignedMessageWithECKey): Boolean

//    /** Revokes a gateway's key from the user.
//     * @param message Message for revocation.
//     * @param signature Signature using user's private key.
//     * @param publicKey Gateway's public key. */
//    fun revokeKey(message: String, signature: String, publicKey: String)
}

